### PR TITLE
cmd: add file lock to work around port conflict

### LINKED ISSUE #4323: TiKV may not report errors if port is in conflict
Because TiKV binds port using SO_REUSEPORT, so if the port is binded by other process with SO_REUSEPORT already, the instance can start successfully without errors. However new connections may not be accepted by the started instance, so it's idle and no errors or warnings are reported.

This can easily happen if multiple instances are sharing the same machine while they are not configured properly.

### PR DESCRIPTION
**Pull request by hunterlxt**
Signed-off-by: TXXT <hunterlxt@live.com>

### What problem does this PR solve?

close #4323 

### What is changed and how it works?

add file lock to work around port conflict

### Related changes

NO

### Check List 

Tests 

- Unit test

Side effects

- Potential Performance regression (Related bench is posted on #4323 )

### Release note 
Add file lock check to avoid starting multiple instances bound to the same address.

### CODE REVIEW

**Comment by siddontang** [comment]
I am so surprised that this can affect performance.

**Comment by hunterlxt** [comment]
> I am so surprised that this can affect performance.

I found this explanation. FYI https://github.com/grpc/grpc/pull/14748

**Comment by siddontang** [comment]
> Clearly, using SO_REUSEPORT gives significant performance improvements (and this is why, it is generally enabled in our code - unless we are running on some older kernel version that didn't have this feature)

So why do we disable this?

**Comment by hunterlxt** [comment]
> > Clearly, using SO_REUSEPORT gives significant performance improvements (and this is why, it is generally enabled in our code - unless we are running on some older kernel version that didn't have this feature)
> 
> So why do we disable this?

Related issue. #4323  By benching in ci to evaluate the impact of performance fallback, if there is an impact, I think that I can find other ways to solve the problem of port conflict

**Comment by hunterlxt** [comment]
@BusyJay PTAL

**Comment by BusyJay** [comment]
You should fill the release note.

**Comment by hunterlxt** [comment]
> You should fill the release note.

I will update if the PR needs be merged

**Comment by hunterlxt** [comment]
I noticed in two ycsb tests, tidb has different hash. @BusyJay . Maybe need more bench

**Comment by siddontang** [comment]
I find most of the bench results are worse than before, right?

**Comment by BusyJay** [comment]
Another approach to get around the issue without introduce potential regression is to lock the port via a file lock. This should be able to avoid port conflict for TiKV instances but can't prevent other processes accidentally conflict. Though it seems enough for the reported cases.

**Comment by hunterlxt** [comment]
@BusyJay I don't know if os will automatically clean up the lock file in `/tmp`, if it is, where you think it is better to place?

File: cmd/src/setup.rs [hunk 1]
```diff
@@ -208,3 +212,18 @@ pub fn validate_and_persist_config(config: &mut TiKvConfig, persist: bool) {
...
+    let file_path = Path::new(LOCK_FILE_BASE).join(format!("{}.tikv.lock", config.server.addr));
+
+    let fd = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .open(file_path)
+        .expect("open lock file failed")
+        .into_raw_fd();
+
+    if unsafe { libc::flock(fd, libc::LOCK_EX | libc::LOCK_NB) } != 0 {
```
Comment (hunterlxt) [inline cmd/src/setup.rs:226]: Does tikv need to support Windows and Mac, and if so do they need similar verification? 
 I will add the platform judgment here later

File: cmd/src/setup.rs [hunk 2]
```diff
@@ -13,6 +15,8 @@ use tikv_util::{self, logger};
 
 // A workaround for checking if log is initialized.
 pub static LOG_INITIALIZED: AtomicBool = AtomicBool::new(false);
+// Lock file directory base
+const LOCK_FILE_BASE: &str = "/tmp";
```
Comment (BusyJay) [inline cmd/src/setup.rs:19]: Should use `std::env::temp_dir`.

File: cmd/src/setup.rs [hunk 3]
```diff
@@ -208,3 +212,18 @@ pub fn validate_and_persist_config(config: &mut TiKvConfig, persist: bool) {
         }
     }
 }
+
+pub fn check_lock_file(config: &TiKvConfig) {
+    let file_path = Path::new(LOCK_FILE_BASE).join(format!("{}.tikv.lock", config.server.addr));
```
Comment (BusyJay) [inline cmd/src/setup.rs:217]: "0.0.0.0:port" should be conflict with all other address.

File: cmd/src/setup.rs (hunk 1 above)
Comment (BusyJay) [reply cmd/src/setup.rs:226]: You can copy the implement of "LOCK" file in tikv-server.

File: cmd/src/server.rs [hunk 4]
```diff
@@ -294,6 +296,53 @@ impl TiKVServer {
...
+        if cur_addr_vec.len() < 2 {
+            fatal!("tikv address has wrong format");
+        }
+        let cur_ip = cur_addr_vec.first().cloned().unwrap().trim();
+        let cur_port = cur_addr_vec.last().cloned().unwrap().trim();
+
+        let search_base = env::temp_dir().join("TIKV_LOCK_FILES");
+        std::fs::create_dir_all(&search_base).expect("create TIKV_LOCK_FILES failed");
+
+        for result in WalkDir::new(&search_base) {
```
Comment (BusyJay) [inline cmd/src/server.rs:311]: Why not just `fs::read_dir`?

File: cmd/src/server.rs [hunk 5]
```diff
@@ -294,6 +296,44 @@ impl TiKVServer {
...
+        let search_base = env::temp_dir().join("TIKV_LOCK_FILES");
+        std::fs::create_dir_all(&search_base).expect("create TIKV_LOCK_FILES failed");
+
+        for result in fs::read_dir(&search_base).unwrap() {
+            if let Ok(entry) = result {
+                if !entry.file_type().unwrap().is_file() {
+                    continue;
+                }
+                let file_path = entry.path();
+                let file_name = file_path.file_name().unwrap().to_str().unwrap().to_owned();
```
Comment (BusyJay) [inline cmd/src/server.rs:318]: Better avoid using colon in file name.

File: cmd/src/server.rs [hunk 6]
```diff
@@ -861,6 +901,24 @@ fn check_system_config(config: &TiKvConfig) {
...
+        fatal!(
+            "failed to create lock at {}: {}",
+            existed_path.as_ref().display(),
+            e
+        )
+    });
+
+    if f.try_lock_exclusive().is_err() {
+        fatal!(
+            "lock {} failed, maybe another instance is binding with this address.",
```
Comment (BusyJay) [inline cmd/src/server.rs:915]: The error message may be confusing.

Comment (hunterlxt) [reply cmd/src/server.rs:915]: I don't know where it is confusing, change to `binding with same port` ?

Comment (hunterlxt) [reply cmd/src/server.rs]: Fixed, and I will add `merged` tag.

**Pull request merged**